modules.d/35network-manager/nm-run.sh

```diff
@@ -24,11 +24,49 @@ if [ -s /run/NetworkManager/initrd/hostname ]; then
     cat /run/NetworkManager/initrd/hostname > /proc/sys/kernel/hostname
 fi
 
+kf_get_string() {
+    # NetworkManager writes keyfiles (glib's GKeyFile API). Have a naive
+    # parser for it.
+    #
+    # But GKeyFile will backslash escape certain keys (\s, \t, \n) but also
+    # escape backslash. As an approximation, interpret the string with printf's
+    # '%b'.
+    #
+    # This is supposed to mimic g_key_file_get_string() (poorly).
+
+    v1="$(sed -n "s/^$1=/=/p" | sed '1!d')"
+    test "$v1" = "${v1#=}" && return 1
+    printf "%b" "${v1#=}"
+}
+
+kf_unescape() {
+    # Another layer of unescaping. While values in GKeyFile format
+    # are backslash escaped, the original strings are backslash escaped
+    # too. This will undo the second layer of escaping.
+    printf "%b" "$1"
+}
+
+printf_q() {
+    # There is no `printf %q` in POSIX shell.
+    bash -c 'printf "%q" "$@"' printf "$@"
```

Comment:
If you need bash, then just change the interpreter of this file to `#!/bin/bash` and let `module-setup.sh` require the `bash` module. Calling `bash` is nonsense :)

Reply:
I had that in version1.

docs/HACKING.md file says `Hooks MUST be POSIX compliant`.

Reply:
Well, not if the module requires the `bash` module.

Reply:
We have several exceptions there.

Reply:
OK.  The HACKING file is the not clear to me.

Because it makes it sound as if the problem is that this script (`nm-run.sh` is a hook, right?) gets sourced by another script, and it's not clear that networkmaanger module depending on bash will make that other script run under bash.

I pushed a fixup. If this is the right way, I'd readd more bashism (e.g.`local` is really nice).